_posts/2024-08-06-quarkus-and-leyden.adoc

```diff
@@ -0,0 +1,275 @@
+
+== Current status of Project Leyden
+
+There are already experimental https://jdk.java.net/leyden/[early-access builds of Leyden] that can be tested based on https://openjdk.org/jeps/8315737[this draft JEP about Ahead-of-Time Class Linking]. With the https://www.youtube.com/watch?v=NlJK5BKXtHI[Leyden Project], the training run idea has been extended to a wider range of data structures embedded in the new AOT cache. Now the training data contains:
+
+ - Class file events with historical data (Classes loaded and linked, Compilations)
+ - Resolution of API points and indy (stored in constant pool images in the AOT archive). If you have lambdas in your code, they are captured here.
+ - Pre-created constant objects in the Java heap (String and Class<?> constants)
+ - Execution profiles and some compiled native code (all tiers)
+
+=== Some known limitations
+
+This is an experimental project being developed by multiple teams having different approaches and focuses. Limitations explained here are being worked on at the time of writing this blog post.
+
+One of the main issues is that functionality is currently only available for x86_64 and AArch64 architectures at the moment. 
+
+Also, current developments rely on a flat classpath. If the application is using custom classloaders, then it may not benefit as much as it could as it may miss caching many classes. 
+
+The same happens if the application is intensively using reflection. Quarkus avoids reflection whenever possible, preferring to resolve reflective calls at build time as well - so there’s a nice synergy right there.
+
+However Quarkus in “fast-jar” mode, which is the default packaging mode, will use a custom classloader which currently would get in the way of some Leyden optimisations. One could use a different packaging mode in Quarkus to get more prominent benefits from Leyden, but doing so would disable other Quarkus optimisations, so the comparison wouldn’t be entirely fair today.
+We hope to work on improvements in this area to have all possible benefits, combined.
+
+The focus on these first early releases has been on bootstrap times. There are measurable, significant startup time improvements, due to AoT loading and linking. In some cases, these improvements on startup time have worsened the memory footprint of some applications. That’s an already known issue that is being worked on, and the expected outcome is to improve memory footprint as well, so we would suggest not worrying too much about total memory consumption at this stage.
+
+Since the AOT archives include machine specific optimisations such as the native code generated by the C2 compiler, the training run and the production run must be done on the same type of hardware and JDK versions; it also requires using the same JAR-based classpaths and the same command line options. 
+
+Although you can use a different Main class for running the application, for example a test class that simulates usage.
```

Comment:
Although the training run can use a different Main class to the one used for running the application, for example a test class that simulates usage.